### ISSUE TITLE
CCNG migrations incompatible with MariaDB 10.3

### ISSUE THREAD

**Issue by sethboyles** [cloudfoundry/cloud_controller_ng#2042]
## Issue

When upgrading the capi-ruby-units Dockerfile to use `ruby-2.7-buster`,  the default mysql server (MariaDB) version was upgraded from 10.1 to 10.3.  This caused our units to fail with the following error: 

```
Sequel::DatabaseError:
  Mysql2::Error: Can't DROP CONSTRAINT `route_mappings_app_guid_route_guid_process_type_key`; check that it exists
```
## Steps to Reproduce

check out this version of the Dockerfile: https://github.com/cloudfoundry/capi-dockerfiles/blob/eb80e58c1233e0b1fc6a26a69e98590472783d33/capi-ruby-units/Dockerfile

Do the following:

1. run `docker build .` 
1. run the image in interactive mode: `docker run -it <image_name> /bin/bash`
1. clone CCNG: `git clone https://github.com/cloudfoundry/cloud_controller_ng.git' 
1. `cd cloud_controller_ng` 
1. start mysql: `service mysql start`
1. run unit tests for MySql: `DB=mysql ./bin/bundle exec rake spec`

## Expected result

Mysql tests pass successfully 

## Current result

```
Sequel::DatabaseError:
  Mysql2::Error: Can't DROP CONSTRAINT `route_mappings_app_guid_route_guid_process_type_key`; check that it exists
```

**Comment by Gerg** [comment]
It looks like the constraint is dropped in this migration: https://github.com/cloudfoundry/cloud_controller_ng/blob/7830ca97b81d657c64fde6bd75f0a399c49a1436/db/migrations/20160601173727_add_port_to_route_mappings.rb#L9

Interestingly, other than the down of that migration, I can't see where that constraint is being added. I would expect to see something like `add_unique_constraint [:app_guid, :route_guid, :process_type]` somewhere in a previous migration (with that constraint name being auto generated by sequel).

My guess is that newer version of MariaDB don't treat dropping constraints idempotently anymore.

**Comment by sethboyles** [comment]
is this the equivalent constraint? https://github.com/cloudfoundry/cloud_controller_ng/blob/main/db/migrations/20160203011824_create_route_mappings_table.rb#L19 

Not sure if the `unique` here differs in some way.

**Comment by Gerg** [comment]
Good eyes: https://github.com/jeremyevans/sequel/blob/master/doc/schema_modification.rdoc#label-unique

Interestingly, the [MariaDB docs](https://mariadb.com/kb/en/alter-table/#drop-constraint) say `DROP CONSTRAINT` for unique constraints weren't added until `10.2.22`/`10.3.13`. Maybe sequel had a different fallback for older versions? Do you know what patch version of MariaDB you used to reproduce this issue?

**Comment by sethboyles** [comment]
```
root@34c5b1254265:/# mariadb --version
mariadb  Ver 15.1 Distrib 10.3.27-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2
```
looks like 10.3.27, so it would have that change.

> Looking at sequel release notes, maybe we need to add a :type => :unique to the drop_constraint call. That doesn't explain why it's working for older MariaDB versions though.

Good find.  Maybe previously was it just silently doing nothing?  Even if that were so, not sure why we wouldn't have seen other issues arise from that constraint still being enforced.

**Comment by sethboyles** [comment]
I was curious as to the state of the DB immediately prior to the `drop_constraint` call, so I pulled this:

```
[11] pry(#<Sequel::Mysql2::Database>)> pp result[:Table][:"Create Table"]
"CREATE TABLE `route_mappings` (\n" +
"  `id` int(11) NOT NULL AUTO_INCREMENT,\n" +
"  `guid` varchar(255) COLLATE utf8_bin NOT NULL,\n" +
"  `created_at` timestamp NOT NULL DEFAULT current_timestamp(),\n" +
"  `updated_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',\n" +
"  `app_guid` varchar(255) COLLATE utf8_bin DEFAULT NULL,\n" +
"  `route_guid` varchar(255) COLLATE utf8_bin DEFAULT NULL,\n" +
"  `process_type` varchar(255) COLLATE utf8_bin DEFAULT NULL,\n" +
"  `app_port` int(11) DEFAULT 8080,\n" +
"  PRIMARY KEY (`id`),\n" +
"  UNIQUE KEY `route_mappings_guid_index` (`guid`),\n" +
"  UNIQUE KEY `app_guid` (`app_guid`,`route_guid`,`process_type`),\n" +
"  KEY `route_mappings_created_at_index` (`created_at`),\n" +
"  KEY `route_mappings_updated_at_index` (`updated_at`),\n" +
"  KEY `route_mappings_app_guid_index` (`app_guid`),\n" +
"  KEY `route_mappings_route_guid_index` (`route_guid`),\n" +
"  KEY `route_mappings_process_type_index` (`process_type`),\n" +
"  CONSTRAINT `route_mappings_ibfk_1` FOREIGN KEY (`app_guid`) REFERENCES `apps_v3` (`guid`),\n" +
"  CONSTRAINT `route_mappings_ibfk_2` FOREIGN KEY (`route_guid`) REFERENCES `routes` (`guid`)\n" +
") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"
```

One thing jumps out at me is that I don't see the full `route_mappings_app_guid_route_guid_process_type_key` name.  It seems to be represented another way.  Maybe it's something like MariaDB (or Sequel?) doesn't automatically form the constraint key in the name we expect?

**Comment by Gerg** [comment]
Maybe that's what it's named in postgres: https://github.com/cloudfoundry/cloud_controller_ng/wiki/CAPI-Migration-Style-Guide#name-all-constraints

**Comment by sethboyles** [comment]
https://github.com/cloudfoundry/cloud_controller_ng/blob/7830ca97b81d657c64fde6bd75f0a399c49a1436/db/migrations/20160601173727_add_port_to_route_mappings.rb#L5-L9

I think this block is an attempt at mitigating the unnamed constraint problem.  Maybe we change it to an if/else, or go back and name the constraint and remove the Mysql check entirely.  I tested out the latter, and it successfully got past the migration... only to fail on another migration:  `https://github.com/cloudfoundry/cloud_controller_ng/blob/main/db/migrations/20171013223336_remove_deprecated_v1_service_fields.rb#L4`

So there may be several more odd issues like this.

**Comment by Gerg** [comment]
I tried running the migrations with an older version of MariaDB (10.2.11), and still saw the same failure. Looks like the changes in 10.2.22 was not the root issue.

**Comment by Gerg** [comment]
This issue was addressed by ee2efad and 93a1464.

**Comment by elenasharma** [comment]
The fix for this issue was released in [capi-release 1.105](https://github.com/cloudfoundry/capi-release/releases/tag/1.105.0). Thank you for raising this issue!
